File: javabot-play/app/controllers/Binders.scala
package controllers

import org.bson.types.ObjectId
import play.api.mvc.Results.Status
import play.api.mvc.{JavascriptLitteral, PathBindable}
import play.api.data.format.Formatter
import play.api.data.FormError
import play.api.http.ContentTypes
import play.api.{Play, PlayException}
import reflect.Manifest


object Binders {

  implicit def bindableOption[T: PathBindable] = new PathBindable[Option[T]] {
    def bind(key: String, value: String) = {
      implicitly[PathBindable[T]].bind(key, value).right.map(Some(_))
    }
    def unbind(key: String, value: Option[T]) = value.map(v => implicitly[PathBindable[T]].unbind(key, v)).getOrElse("")
  }

  implicit def bindableObjectId = new PathBindable[ObjectId] {
    def bind(key: String, value: String) = {
      if (ObjectId.isValid(value)) {
        Right(new ObjectId(value))
      } else {
        Left("Cannot parse parameter " + key + " as BSON ObjectId")
      }
    }

    def unbind(key: String, value: ObjectId) = value.toString
  }

  implicit def bindableJavascriptLiteral = new JavascriptLitteral[ObjectId] {
    def to(value: ObjectId) = value.toString
  }

}

File: web/app/controllers/Binders.java
package controllers;

import org.bson.types.ObjectId;
import play.api.mvc.PathBindable;
import scala.Function1;
import scala.util.Either;
import scala.util.Left;
import scala.util.Right;

public class Binders {
  public static PathBindable<ObjectId> objectIdBinder = new PathBindable<ObjectId>() {
    @Override
    public Either<String, ObjectId> bind(final String key, final String value) {
      return (ObjectId.isValid(value))
          ? new Right(new ObjectId(value))
          : new Left("Cannot parse parameter " + key + " as BSON ObjectId");
    }

    @Override
    public String unbind(final String key, final ObjectId value) {
      return value.toString();
    }

    @Override
    public String javascriptUnbind() {
      return null;
    }

    @Override
    public <B> Object transform(final Function1<ObjectId, B> toB, final Function1<B, ObjectId> toA) {
      return null;
    }
  };
}
